### PR TITLE
IPv6 support for endpoint.py

### PR DESCRIPTION
**Pull request by glyph**
Similar to https://github.com/matrix-org/synapse/pull/1689, but for endpoint.py

### CODE REVIEW

File: synapse/http/endpoint.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from twisted.internet.endpoints import SSL4ClientEndpoint, TCP4ClientEndpoint
+from twisted.internet.endpoints import HostnameEndpoint, wrapClientTLS
```
Comment (kyrias) [inline synapse/http/endpoint.py]: `TCP4ClientEndpoint` is used as a default value in in the `SpiderEndpoint` and `SRVClientEndpoint` init functions.

File: synapse/http/endpoint.py [hunk 2]
```diff
@@ -58,11 +58,11 @@ def matrix_federation_endpoint(reactor, destination, ssl_context_factory=None,
...
 
     if ssl_context_factory is None:
-        transport_endpoint = TCP4ClientEndpoint
+        transport_endpoint = HostnameEndpoint
         default_port = 8008
     else:
-        transport_endpoint = SSL4ClientEndpoint
-        endpoint_kw_args.update(sslContextFactory=ssl_context_factory)
+        def transport_endpoint(reactor, host, port):
+            return wrapClientTLS(ssl_context_factory, HostnameEndpoint(reactor, host, port))
```
Comment (kyrias) [inline synapse/http/endpoint.py]: This wrapper also needs to take a timeout parameter for use in HostnameEndpoint.

**Comment by kyrias** [comment]
Looks good other than those small things, and I've tested it and it seems to be working!

**Comment by kyrias** [comment]
Federation seems to be at least partially broken over IPv6 though.  I can join rooms and send/accept invites over IPv6, but messages aren't being synced.

File: synapse/http/endpoint.py (hunk 1 above)
Comment (glyph) [inline synapse/http/endpoint.py]: It should probably be removed from there; `TCP4ClientEndpoint` is only interesting as a low-level implementation detail.

**Comment by glyph** [comment]
Since this has been superseded, shall I close it?

File: synapse/http/endpoint.py (hunk 1 above)
Comment (glyph) [inline synapse/http/endpoint.py]: Gotcha; yeah.  I'd be inclined to eliminate the default, personally; I find that default args are useful a lot more rarely than they _seem_ like they'd be useful :)

**Comment by ara4n** [comment]
@kyrias has consolidated this into #1696, and we'll try to land it from there. thanks @glyph!!

**Pull request closed without merging**